### ISSUE TITLE
我尝试了concat的方式来实现这个结构，但是效果不太好，麻烦可以看下存在哪些问题吗

### ISSUE THREAD

**Issue by bestzdy** [INVOKERer/DeepRFT#65]
<img width="674" height="823" alt="Image" src="https://github.com/user-attachments/assets/7e3beca5-e09b-4340-aee0-e9fb900dffde" />

```markdown

```使用concat形式的Res-FFT-RELU-Block，参考论文补充材料中的结构：
class NAFBlock(nn.Module):
    def __init__(self, c, DW_Expand=2, FFN_Expand=2, drop_out_rate=0.):
        super().__init__()
        dw_channel = c * DW_Expand
        self.conv1 = nn.Conv2d(in_channels=c, out_channels=dw_channel, kernel_size=1, padding=0, stride=1, groups=1, bias=True)
        self.conv2 = nn.Conv2d(in_channels=dw_channel, out_channels=dw_channel, kernel_size=3, padding=1, stride=1, groups=dw_channel,
                               bias=True)
        self.conv3 = nn.Conv2d(in_channels=dw_channel // 2, out_channels=c, kernel_size=1, padding=0, stride=1, groups=1, bias=True)
        
        # Simplified Channel Attention
        self.sca = nn.Sequential(
            nn.AdaptiveAvgPool2d(1),
            nn.Conv2d(in_channels=dw_channel // 2, out_channels=dw_channel // 2, kernel_size=1, padding=0, stride=1,
                      groups=1, bias=True),
        )

        # SimpleGate
        self.sg = SimpleGate()

        ffn_channel = FFN_Expand * c
        self.conv4 = nn.Conv2d(in_channels=c, out_channels=ffn_channel, kernel_size=1, padding=0, stride=1, groups=1, bias=True)
        self.conv5 = nn.Conv2d(in_channels=ffn_channel // 2, out_channels=c, kernel_size=1, padding=0, stride=1, groups=1, bias=True)

        self.norm1 = LayerNorm2d(c)
        self.norm2 = LayerNorm2d(c)

        self.dropout1 = nn.Dropout(drop_out_rate) if drop_out_rate > 0. else nn.Identity()
        self.dropout2 = nn.Dropout(drop_out_rate) if drop_out_rate > 0. else nn.Identity()

        self.beta = nn.Parameter(torch.zeros((1, c, 1, 1)), requires_grad=True)
        self.gamma = nn.Parameter(torch.zeros((1, c, 1, 1)), requires_grad=True)  
# 上面是原本的NAFNet不动，下面是1*1conv,GELU,1*1conv
        if self.usefft:
            self.fftconv1 = nn.Sequential(
            nn.Conv2d(in_channels=c*2, out_channels=c*2, kernel_size=1, padding=0, stride=1, groups=1, bias=True),
            nn.GELU(),
            nn.Conv2d(in_channels=c*2, out_channels=c*2, kernel_size=1, padding=0, stride=1, groups=1, bias=True)
            )
            self.fftconv2 = nn.Sequential(
            nn.Conv2d(in_channels=c*2, out_channels=c*2, kernel_size=1, padding=0, stride=1, groups=1, bias=True),
            nn.GELU(),
            nn.Conv2d(in_channels=c*2, out_channels=c*2, kernel_size=1, padding=0, stride=1, groups=1, bias=True)
            )
#------------------------------------------------------------
    def forward(self, inp):
        x= inp
        x1 = self.norm1(x)
        x = self.conv1(x1)
        x = self.conv2(x)
        x = self.sg(x)
        x = x * self.sca(x)
        x = self.conv3(x)

        x = self.dropout1(x)
        y = inp + x * self.beta
# 前半部分原本的不动，额外加一个FFT分支
        if self.usefft:
            xf = torch.fft.rfft2(x1, norm="backward")# <---here，前半部分
            xf_imag = xf.imag
            xf_real = xf.real
            yf = torch.cat([xf_real, xf_imag], dim=1)
            ffty = self.fftconv1(yf)
            y_real, y_imag = torch.chunk(ffty, 2, dim=1)
            ffty = torch.complex(y_real, y_imag)
            fft_y = torch.fft.irfft2(ffty, norm="backward")
            y = y + fft_y * self.beta 

        x2 = self.norm2(y)
        x = self.conv4(x2)
        x = self.sg(x)
        x = self.conv5(x)

        x = self.dropout2(x)
        out = y + x * self.gamma
# 后半部分原本的不动，额外加一个FFT分支
        if self.usefft:
            xf = torch.fft.rfft2(x2, norm="backward")# <--here，后半部分
            xf_imag = xf.imag
            xf_real = xf.real
            yf = torch.cat([xf_real, xf_imag], dim=1)
            ffty = self.fftconv2(yf)
            y_real, y_imag = torch.chunk(ffty, 2, dim=1)
            ffty = torch.complex(y_real, y_imag)
            fft_y = torch.fft.irfft2(ffty, norm="backward")
            out = out + fft_y * self.gamma

        return  out

**Comment by INVOKERer** [comment]
你好，你的训练过程中的loss下降是否正常，测试是否有使用sliding window的方式进行？

**Comment by bestzdy** [comment]
loss下降明显变慢了，使用的是CharbonnierLoss+fft_loss，同iter大概损失高0.005。batchsize设置的是8，optimizer与scheduler的设置与NAFNet一致。测试时是256大小的window对原图切块再拼回，这个是sliding window吗。

**Comment by INVOKERer** [comment]
这个项目可能有点太老了，FNAFBlock和sliding window的方式可以参考AdaRevD项目中的代码

**Comment by bestzdy** [comment]
好的，谢谢了
